accepted/single-file/design.md

```diff
@@ -66,225 +57,211 @@ Publishing to a single file can be triggered by adding the following property to
+### Startup
 
-### New API
+On Startup, the [host components](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-components.md) perform the following functions:
 
-In this section, we propose adding a few APIs to facilitate common operations on bundled-apps.
+* **AppHost**: The AppHost identifies itself as a single-file bundle (by checking the [bundle marker](bundler.md#bundle-marker)) before invoking [HostFxr](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-components.md#host-fxr).
 
-The binaries that are published in the project are expected to be handled transparently by the host. However, explicit access to the embedded files is useful in situations such as:
+* **HostFxr**: If invoked from a single-file app, HostFxr process the `runtimeconfig.json` and `deps.json` files directly from the bundle. The location of these `json` files  are identified directly from the bundle-header for simple access.
 
-- Reading additional files packaged into the app (ex: data files).
-- Open an assembly for reflection/inspection
-- Load plugins built as single-file class-libs using existing Loader APIs.
+* **HostPolicy**: Much of the bundle-processing is performed within HostPolicy.
+  
+  * If the app needs extraction, extracts out the appropriate files as explained in [this document](extractor.md).
+  
+  * Reads the `deps.json` file directly from the bundle and resolves dependencies.
+  
+  * Processes the Bundle manifest to maintain internal data-structures to locate bundled assemblies when probed by the runtime.
+  
+  * Implements a bundle-probe function, which will be used by the runtime to probe contents of the bundle when resolving assemblies. 
+  
+    ```C++
+  /// <summary>
+    /// <param name="path"> Relative-path to the file being probed. </param>
+    /// <param name="size"> Out-param: size of the file, if found. </param>
+    /// <param name="path"> Out-param: offset within the bundle, if found</param>
+    /// <returns> true if the requested file is found in the bundle, 
+    ///           false otherwise. </returns>
+    /// </summary>
+    bool bundle_probe(const char *path, int64_t *size, int64_t *offset);
+    ```
+    
+  * The `bundle_probe` function pointer is passed to the runtime (encoded as a string) through a property named `BUNDLE_PROBE`.
+  
+  * When probing for assemblies, the the [assembly resolution logic](https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/docs/design/features/assembly-conflict-resolution.md#probe-ordering) will treat bundled assemblies similar to assemblies in the app directory. The probe ordering will be in the order:  servicing location, shared store, framework directory(s) from higher to lower, the *single-file bundle*, app directory, additional locations specified in runtimeconfig.dev.json file.
```

Comment:
@vitek-karas  it is interesting whether the "App directory" should be a part of the default probing paths at all, for single-files. 

If the user chooses to not bundle certain assemblies, those files will be left beside the app in the publish directory. We'd like to be able to load these assemblies without having to set `additionalProbingPaths` 

However, the single file app may live in a directory with many other apps, assemblies and random files. The app shouldn't pick up a different assembly simply because it lies beside the app and has a later version.

Reply:
The probing doesn't work that way. The algorithm goes in the order of probing paths and if it finds a match, it stops and returns. The host never opens the file (it only checks for existence), and so it doesn't read the file version (or assembly version).

There is some logic comparing versions, but it's solely based on versions written in `.deps.json` - and it only works to resolve conflicts between the app and the frameworks.

Reply:
Also forgot to add: I do think we should probe the app's path for files which are left beside it (and for cases where the app uses plugins and such).
There's an interesting question what should happen for apps which are single-file and extracted. In that case the probing should probably look both in the extraction folder and in the app's .exe folder (in that order).

Reply:
Thanks for the clarification @vitek-karas, I forgot about this detail -- we'd talked about this issue.